src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -1763,7 +1749,16 @@ public void ReturnHttp11Connection(HttpConnection connection, bool isNewConnecti
                         isNewConnection = false;
                     }
 
-                    if (_http11RequestQueue.TryDequeueWaiter(out waiter))
+                    if (initialRequestWaiter is not null)
+                    {
+                        // Try to handle the request that we initiated the connection for first
+                        waiter = initialRequestWaiter;
+                        initialRequestWaiter = null;
+
+                        // Remove the initial request if it is currently at the head of the queue
+                        _http11RequestQueue.TryDequeueSpecificWaiter(waiter);
```

Comment:
Can you add a comment about why we ignore the return value?  Presumably it's because if this returns true, we own the waiter and expect the waiter.TrySetResult logic later to succeed, and if this returns false, we rely on that TrySetResult returning false and then us looping around again?

Reply:
We want the behavior of `ReturnHttp11Connection` to be such that if we have found a request to service, that request is removed if it was at the head of the queue.

Normally, this method would call into `TryDequeueWaiter` which will do the removal.
For the initial request case, `TryDequeueSpecificWaiter` matches that behavior.
Alternatively, we could call `PruneCompletedRequests` at the end of the method, but that would force us to take the lock again.

There are other places that would have eventually removed the entry from the queue, but there is one specific scenario where doing the remove here is a must:
- The user has the `MaxConnectionsPerServer` set
- The user does not dispose `HttpResponseMessage`s and relies on the GC to finalize it and eventually signal to the pool that a connection slot is available

If we did not remove the initial request here, the pool would keep all of the responses rooted (via a reference to the `HttpConnection` in the waiter).
With this removal, we guarantee that at least one response will eventually be finalized and prune the queue.

Reply:
Thanks.  That'd be a great comment :)

Reply:
If this method returned `true`, the `TrySetResult` could still fail if the request was already canceled.
If it returned `false`, the request could be in any state (waiting, already processed or canceled), just not at the head of the queue.

Either way, there's nothing that the caller logic could do with the return value.